routing/routing_tests/restriction_test.cpp

```diff
@@ -0,0 +1,768 @@
+           false /* graphWithoutRestrictions */, *m_graph);
+  ApplyRestrictionNoRealFeatures(RestrictionPoint({0 /* feature id */, 1 /* point id */},
+                                 {3 /* feature id */, 0 /* point id */}, restictionCenterId));
+  EdgeTest(restictionCenterId, 2 /* expectedIntgoingNum */, 3 /* expectedOutgoingNum */,
+           false /* graphWithoutRestrictions */, *m_graph);
+  EdgeTest(restictionCenterId, 3 /* expectedIntgoingNum */, 3 /* expectedOutgoingNum */,
+           true /* graphWithoutRestrictions */, *m_graph);
+
+  // Testing route building after adding the restriction.
+  SetStarter(RoadPoint(0, 0) /* start */, RoadPoint(5, 0) /* finish */);
+  vector<m2::PointD> const expectedGeom = {
+      {2 /* x */, 0 /* y */}, {1, 0}, {0, 0}, {0, 1}, {0.5, 1}};
+  TestRouteGeometry(*m_starter, AStarAlgorithm<IndexGraphStarter>::Result::OK, expectedGeom);
+}
+
+// Route through flag graph with one restriciton (type only) from F0 to F1.
+UNIT_CLASS_TEST(RestrictionTest, FlagGraph_RestrictionF0F1Only)
+{
+  Init(BuildFlagGraph());
+  Joint::Id const restictionCenterId = GetJointIdForTesting({0, 1});
+  ApplyRestrictionOnlyRealFeatures(RestrictionPoint({0 /* feature id */, 1 /* point id */},
+                                   {1 /* feature id */, 0 /* point id */}, restictionCenterId));
+
+  SetStarter(RoadPoint(0, 0) /* start */, RoadPoint(5, 0) /* finish */);
+  vector<RoadPoint> const expectedRoute = {{IndexGraph::kStartFakeFeatureIds, 0 /* point id */},
+                                           {IndexGraph::kStartFakeFeatureIds, 1},
+                                           {IndexGraph::kStartFakeFeatureIds + 1, 1},
+                                           {2 /* feature id */, 1 /* point id */},
+                                           {4, 1}};
+  TestRouteSegments(*m_starter, AStarAlgorithm<IndexGraphStarter>::Result::OK, expectedRoute);
+}
+
+UNIT_CLASS_TEST(RestrictionTest, FlagGraph_PermutationsF1F3NoF7F8OnlyF8F4OnlyF4F6Only)
+{
+  Init(BuildFlagGraph());
+  RestrictionVec const restrictions = {{Restriction::Type::No, {0 /* feature from */, 3 /* feature to */}},
+                                       {Restriction::Type::Only, {0 /* feature from */, 1 /* feature to */}},
+                                       {Restriction::Type::Only, {1 /* feature from */, 2 /* feature to */}}};
+
+  vector<m2::PointD> const expectedGeom = {{2 /* x */, 0 /* y */}, {1, 0}, {0, 0}, {0, 1}, {0.5, 1}};
+  TestRestrictionPermutations(restrictions, expectedGeom, RoadPoint(0, 0) /* start */,
+                              RoadPoint(5, 0) /* finish */, *this);
+}
+
+// 1 *-F4-*-F5-*---F6---* Finish
+//   |         |
+//   F2        F3
+//   |         |
+// 0 *---F1----*---F0---* Start
+//   0         1        2
+// Note 1. All features are two-way. (It's possible to move along any direction of the features.)
+// Note 2. Any feature contains of one segment.
+unique_ptr<IndexGraph> BuildPosterGraph()
```

Comment:
I named all the graph to address to them in comments. 
Poster is a large printed picture used for decoration with two handles but rotate with 90 degrees :)
Have a look at the picture above.